File: volta/test/test_lexer.ml

open Volta.Lexer
module T = Volta.Token_type 
module Lexer = Volta.Lexer
open OUnit2

(* Helper function to create a token for testing *)
let make_token ttype lexeme line column =
  { Lexer.ttype; lexeme; line; column }

(* Helper function to test lexing a single token *)
let test_single_token input expected_token _test_ctxt =
  let lexer = Lexer.create input in
  let token = Lexer.scan_token lexer in
  assert_equal expected_token token
    ~printer:(fun t -> Printf.sprintf "{ ttype = %s; lexeme = \"%s\"; line = %d; column = %d }"
      (T.token_to_string t.ttype) t.lexeme t.line t.column)

(* Helper function to test tokenizing an entire string *)
let test_tokenize input expected_tokens _test_ctxt =
  let lexer = Lexer.create input in
  let tokens = Lexer.tokenize lexer in
  assert_equal expected_tokens tokens
    ~printer:(fun tokens ->
      String.concat "; " (List.map (fun t -> T.token_to_string t.ttype) tokens))

(* Keywords tests *)
let keyword_tests = "Keyword Tests" >::: [
  "test fn keyword" >:: test_single_token "fn" (make_token T.Function "fn" 1 1);
  "test return keyword" >:: test_single_token "return" (make_token T.Return "return" 1 1);
  "test if keyword" >:: test_single_token "if" (make_token T.If "if" 1 1);
  "test else keyword" >:: test_single_token "else" (make_token T.Else "else" 1 1);
  "test while keyword" >:: test_single_token "while" (make_token T.While "while" 1 1);
  "test for keyword" >:: test_single_token "for" (make_token T.For "for" 1 1);
  "test in keyword" >:: test_single_token "in" (make_token T.In "in" 1 1);
  "test match keyword" >:: test_single_token "match" (make_token T.Match "match" 1 1);
  "test struct keyword" >:: test_single_token "struct" (make_token T.Struct "struct" 1 1);
  "test mut keyword" >:: test_single_token "mut" (make_token T.Mut "mut" 1 1);
  "test some keyword" >:: test_single_token "some" (make_token T.SomeKeyword "some" 1 1);
  "test none keyword" >:: test_single_token "none" (make_token T.NoneKeyword "none" 1 1);
]

(* Operator tests *)
let operator_tests = "Operator Tests" >::: [
  (* Arithmetic operators *)
  "test plus" >:: test_single_token "+" (make_token Plus "+" 1 1);
  "test minus" >:: test_single_token "-" (make_token Minus "-" 1 1);
  "test mult" >:: test_single_token "*" (make_token Mult "*" 1 1);
  "test div" >:: test_single_token "/" (make_token Div "/" 1 1);
  "test modulo" >:: test_single_token "%" (make_token Modulo "%" 1 1);
  "test power" >:: test_single_token "**" (make_token Power "**" 1 1);

  (* Comparison operators *)
  "test equals" >:: test_single_token "==" (make_token Equals "==" 1 1);
  "test not equals" >:: test_single_token "!=" (make_token NotEquals "!=" 1 1);
  "test less than" >:: test_single_token "<" (make_token LT "<" 1 1);
  "test greater than" >:: test_single_token ">" (make_token GT ">" 1 1);
  "test less equal" >:: test_single_token "<=" (make_token LEQ "<=" 1 1);
  "test greater equal" >:: test_single_token ">=" (make_token GEQ ">=" 1 1);

  (* Assignment operators *)
  "test assign" >:: test_single_token "=" (make_token Assign "=" 1 1);
  "test plus assign" >:: test_single_token "+=" (make_token PlusAssign "+=" 1 1);
  "test minus assign" >:: test_single_token "-=" (make_token MinusAssign "-=" 1 1);
  "test mult assign" >:: test_single_token "*=" (make_token MultAssign "*=" 1 1);
  "test div assign" >:: test_single_token "/=" (make_token DivAssign "/=" 1 1);
  "test infer assign" >:: test_single_token ":=" (make_token InferAssign ":=" 1 1);
]

(* Delimiter tests *)
let delimiter_tests = "Delimiter Tests" >::: [
  "test left paren" >:: test_single_token "(" (make_token LParen "(" 1 1);
  "test right paren" >:: test_single_token ")" (make_token RParen ")" 1 1);
  "test left brace" >:: test_single_token "{" (make_token LBrace "{" 1 1);
  "test right brace" >:: test_single_token "}" (make_token RBrace "}" 1 1);
  "test left bracket" >:: test_single_token "[" (make_token LSquare "[" 1 1);
  "test right bracket" >:: test_single_token "]" (make_token RSquare "]" 1 1);
  "test colon" >:: test_single_token ":" (make_token Colon ":" 1 1);
]

(* Literal tests *)
let literal_tests = "Literal Tests" >::: [
  (* Integer literals *)
  "test int 0" >:: test_single_token "0" (make_token Integer "0" 1 1);
  "test int 42" >:: test_single_token "42" (make_token Integer "42" 1 1);
  "test int 123" >:: test_single_token "123" (make_token Integer "123" 1 1);

  (* Float literals *)
  "test float 3.14" >:: test_single_token "3.14" (make_token Real "3.14" 1 1);
  "test float 0.5" >:: test_single_token "0.5" (make_token Real "0.5" 1 1);
  "test float 2.0" >:: test_single_token "2.0" (make_token Real "2.0" 1 1);

  (* String literals *)
  "test empty string" >:: test_single_token "\"\"" (make_token StringLiteral "" 1 1);
  "test simple string" >:: test_single_token "\"hello\"" (make_token StringLiteral "hello" 1 1);
  "test string with spaces" >:: test_single_token "\"hello world\"" (make_token StringLiteral "hello world" 1 1);
  "test string with escape" >:: test_single_token "\"hello\\nworld\"" (make_token StringLiteral "hello\nworld" 1 1);
]

(* Identifier tests *)
let identifier_tests = "Identifier Tests" >::: [
  "test simple identifier" >:: test_single_token "x" (make_token Identifier "x" 1 1);
  "test identifier with underscore" >:: test_single_token "my_var" (make_token Identifier "my_var" 1 1);
  "test identifier with numbers" >:: test_single_token "var123" (make_token Identifier "var123" 1 1);
  "test identifier starting with underscore" >:: test_single_token "_private" (make_token Identifier "_private" 1 1);
  "test camelCase identifier" >:: test_single_token "camelCase" (make_token Identifier "camelCase" 1 1);
]

(* Integration tests - testing multiple tokens *)
let integration_tests = "Integration Tests" >::: [
  "test simple expression" >:: test_tokenize "1 + 2" [
    make_token Integer "1" 1 1;
    make_token Plus "+" 1 3;
    make_token Integer "2" 1 5;
    make_token Eof "" 1 6
  ];

  "test variable assignment" >:: test_tokenize "x = 42" [
    make_token Identifier "x" 1 1;
    make_token Assign "=" 1 3;
    make_token Integer "42" 1 5;
    make_token Eof "" 1 7
  ];

  "test function call" >:: test_tokenize "add(1, 2)" [
    make_token Identifier "add" 1 1;
    make_token LParen "(" 1 4;
    make_token Integer "1" 1 5;
    make_token Comma "," 1 6;
    make_token Integer "2" 1 8;
    make_token RParen ")" 1 9;
    make_token Eof "" 1 10
  ];
]

(* Whitespace handling tests *)
let whitespace_tests = "Whitespace Tests" >::: [
  "test spaces between tokens" >:: test_tokenize "1 + 2" [
    make_token Integer "1" 1 1;
    make_token Plus "+" 1 3;
    make_token Integer "2" 1 5;
    make_token Eof "" 1 6
  ];

  "test multiple spaces" >:: test_tokenize "1   +   2" [
    make_token Integer "1" 1 1;
    make_token Plus "+" 1 5;
    make_token Integer "2" 1 9;
    make_token Eof "" 1 10
  ];
]

(* Main test suite *)
let suite = "Volta Lexer Test Suite" >::: [
  keyword_tests;
  operator_tests;
  delimiter_tests;
  literal_tests;
  identifier_tests;
  integration_tests;
  whitespace_tests;
]

(* Run the tests *)
let () = run_test_tt_main suite 

File: tests/test_lexer.cpp
#include "lexer/lexer.hpp"
#include "lexer/token.hpp"
#include <gtest/gtest.h>

using namespace volta::lexer;

// Test fixture for Lexer tests
class LexerTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

TEST_F(LexerTest, SimpleTokens) {
    Lexer lexer("1 + 2");
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 4);
    EXPECT_EQ(tokens[0].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[0].lexeme, "1");
    EXPECT_EQ(tokens[0].line, 1);
    EXPECT_EQ(tokens[0].column, 1);

    EXPECT_EQ(tokens[1].type, TokenType::PLUS);
    EXPECT_EQ(tokens[1].lexeme, "+");

    EXPECT_EQ(tokens[2].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[2].lexeme, "2");

    EXPECT_EQ(tokens[3].type, TokenType::END_OF_FILE);
}

TEST_F(LexerTest, Keywords) {
    Lexer lexer("fn if else while for return");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 7);
    EXPECT_EQ(tokens[0].type, TokenType::FUNCTION);
    EXPECT_EQ(tokens[1].type, TokenType::IF);
    EXPECT_EQ(tokens[2].type, TokenType::ELSE);
    EXPECT_EQ(tokens[3].type, TokenType::WHILE);
    EXPECT_EQ(tokens[4].type, TokenType::FOR);
    EXPECT_EQ(tokens[5].type, TokenType::RETURN);
    EXPECT_EQ(tokens[6].type, TokenType::END_OF_FILE);
}

TEST_F(LexerTest, Identifiers) {
    Lexer lexer("x my_var camelCase _private");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 5);
    EXPECT_EQ(tokens[0].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[0].lexeme, "x");
    EXPECT_EQ(tokens[1].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[1].lexeme, "my_var");
    EXPECT_EQ(tokens[2].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[2].lexeme, "camelCase");
    EXPECT_EQ(tokens[3].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[3].lexeme, "_private");
}

TEST_F(LexerTest, Strings) {
    Lexer lexer("\"hello world\" \"test\\n\" \"quote: \\\"hi\\\"\"");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 4);
    EXPECT_EQ(tokens[0].type, TokenType::STRING_LITERAL);
    EXPECT_EQ(tokens[0].lexeme, "hello world");
    EXPECT_EQ(tokens[1].type, TokenType::STRING_LITERAL);
    EXPECT_EQ(tokens[1].lexeme, "test\n");
    EXPECT_EQ(tokens[2].type, TokenType::STRING_LITERAL);
    EXPECT_EQ(tokens[2].lexeme, "quote: \"hi\"");
}

TEST_F(LexerTest, Integers) {
    Lexer lexer("0 42 123 999");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 5);
    EXPECT_EQ(tokens[0].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[0].lexeme, "0");
    EXPECT_EQ(tokens[1].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[1].lexeme, "42");
    EXPECT_EQ(tokens[2].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[2].lexeme, "123");
    EXPECT_EQ(tokens[3].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[3].lexeme, "999");
}

TEST_F(LexerTest, RealNumbers) {
    Lexer lexer("3.14 0.5 2.0");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 4);
    EXPECT_EQ(tokens[0].type, TokenType::REAL);
    EXPECT_EQ(tokens[0].lexeme, "3.14");
    EXPECT_EQ(tokens[1].type, TokenType::REAL);
    EXPECT_EQ(tokens[1].lexeme, "0.5");
    EXPECT_EQ(tokens[2].type, TokenType::REAL);
    EXPECT_EQ(tokens[2].lexeme, "2.0");
}

TEST_F(LexerTest, ScientificNotation) {
    Lexer lexer("1.5e10 2e-3 3.0e+5");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 4);
    EXPECT_EQ(tokens[0].type, TokenType::REAL);
    EXPECT_EQ(tokens[0].lexeme, "1.5e10");
    EXPECT_EQ(tokens[1].type, TokenType::REAL);
    EXPECT_EQ(tokens[1].lexeme, "2e-3");
    EXPECT_EQ(tokens[2].type, TokenType::REAL);
    EXPECT_EQ(tokens[2].lexeme, "3.0e+5");
}

TEST_F(LexerTest, Operators) {
    Lexer lexer("+ - * / % ** == != < > <= >=");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 13);
    EXPECT_EQ(tokens[0].type, TokenType::PLUS);
    EXPECT_EQ(tokens[1].type, TokenType::MINUS);
    EXPECT_EQ(tokens[2].type, TokenType::MULT);
    EXPECT_EQ(tokens[3].type, TokenType::DIV);
    EXPECT_EQ(tokens[4].type, TokenType::MODULO);
    EXPECT_EQ(tokens[5].type, TokenType::POWER);
    EXPECT_EQ(tokens[6].type, TokenType::EQUALS);
    EXPECT_EQ(tokens[7].type, TokenType::NOT_EQUALS);
    EXPECT_EQ(tokens[8].type, TokenType::LT);
    EXPECT_EQ(tokens[9].type, TokenType::GT);
    EXPECT_EQ(tokens[10].type, TokenType::LEQ);
    EXPECT_EQ(tokens[11].type, TokenType::GEQ);
}

TEST_F(LexerTest, AssignmentOperators) {
    Lexer lexer("= := += -= *= /=");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 7);
    EXPECT_EQ(tokens[0].type, TokenType::ASSIGN);
    EXPECT_EQ(tokens[1].type, TokenType::INFER_ASSIGN);
    EXPECT_EQ(tokens[2].type, TokenType::PLUS_ASSIGN);
    EXPECT_EQ(tokens[3].type, TokenType::MINUS_ASSIGN);
    EXPECT_EQ(tokens[4].type, TokenType::MULT_ASSIGN);
    EXPECT_EQ(tokens[5].type, TokenType::DIV_ASSIGN);
}

TEST_F(LexerTest, Delimiters) {
    Lexer lexer("( ) [ ] { } , ; : .");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 11);
    EXPECT_EQ(tokens[0].type, TokenType::LPAREN);
    EXPECT_EQ(tokens[1].type, TokenType::RPAREN);
    EXPECT_EQ(tokens[2].type, TokenType::LSQUARE);
    EXPECT_EQ(tokens[3].type, TokenType::RSQUARE);
    EXPECT_EQ(tokens[4].type, TokenType::LBRACE);
    EXPECT_EQ(tokens[5].type, TokenType::RBRACE);
    EXPECT_EQ(tokens[6].type, TokenType::COMMA);
    EXPECT_EQ(tokens[7].type, TokenType::SEMICOLON);
    EXPECT_EQ(tokens[8].type, TokenType::COLON);
    EXPECT_EQ(tokens[9].type, TokenType::DOT);
}

TEST_F(LexerTest, Arrows) {
    Lexer lexer("-> =>");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 3);
    EXPECT_EQ(tokens[0].type, TokenType::ARROW);
    EXPECT_EQ(tokens[0].lexeme, "->");
    EXPECT_EQ(tokens[1].type, TokenType::MATCH_ARROW);
    EXPECT_EQ(tokens[1].lexeme, "=>");
}

TEST_F(LexerTest, Comments) {
    Lexer lexer("x := 42 # this is a comment\ny := 10");
    auto tokens = lexer.tokenize();

    ASSERT_EQ(tokens.size(), 7); // x, :=, 42, y, :=, 10, EOF
    EXPECT_EQ(tokens[0].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[0].lexeme, "x");
    EXPECT_EQ(tokens[1].type, TokenType::INFER_ASSIGN);
    EXPECT_EQ(tokens[2].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[2].lexeme, "42");
    EXPECT_EQ(tokens[3].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[3].lexeme, "y");
}

TEST_F(LexerTest, LineAndColumnTracking) {
    Lexer lexer("x := 42\ny := 10");
    auto tokens = lexer.tokenize();

    EXPECT_EQ(tokens[0].line, 1);
    EXPECT_EQ(tokens[0].column, 1);
    EXPECT_EQ(tokens[1].line, 1);
    EXPECT_EQ(tokens[1].column, 3);
    EXPECT_EQ(tokens[3].line, 2);
    EXPECT_EQ(tokens[3].column, 1);
}

TEST_F(LexerTest, ComplexExpression) {
    Lexer lexer("fn add(a: int, b: int) -> int { return a + b }");
    auto tokens = lexer.tokenize();

    EXPECT_EQ(tokens[0].type, TokenType::FUNCTION);
    EXPECT_EQ(tokens[1].type, TokenType::IDENTIFIER);
    EXPECT_EQ(tokens[1].lexeme, "add");
    EXPECT_EQ(tokens[2].type, TokenType::LPAREN);
    EXPECT_EQ(tokens.back().type, TokenType::END_OF_FILE);
}

TEST_F(LexerTest, UnterminatedString) {
    Lexer lexer("\"hello");
    auto tokens = lexer.tokenize();

    EXPECT_TRUE(lexer.getErrorReporter().hasErrors());
    EXPECT_EQ(lexer.getErrorReporter().getErrorCount(), 1);
}

TEST_F(LexerTest, InvalidScientificNotation) {
    Lexer lexer("1e");
    auto tokens = lexer.tokenize();

    EXPECT_TRUE(lexer.getErrorReporter().hasErrors());
    EXPECT_EQ(lexer.getErrorReporter().getErrorCount(), 1);
}